### ISSUE TITLE
log-pane command takes no effect after git pull to ef8105daa3

### ISSUE THREAD

**Issue by wangzhaohe** [leo-editor/leo-editor#4123]
None of the expand/contract-*-pane commands work with the new layout code.

<details>

Leo Log Window
Leo 6.8.2-devel, devel branch, build ef8105daa3
2024-10-22 11:02:35 -0500
Python 3.12.0, PyQt version 6.5.3
darwin

3 command takes no effect after git pull to ef8105daa3

    hide-log-pane
    expand-log-pane
    contract-log-pane

<details>

**Comment by edreamleo** [comment]
@tbpassin @wangzhaohe Thanks for this report. None of these commands work for me on Windows.

**Comment by tbpassin** [comment]
When I wrote that these commands worked for me, I was using the "vertical-thirds" layout. They may work differently with different layouts.

I think these commands assume that the log pane shares the secondary splitter with some other widget, like the outline frame. Some of the new layouts may not meet this assumption. Or the commands may act differently for different layouts.

For example, the "contract-log-pane" command will do nothing if the secondary splitter contains more or fewer than two child widgets. That's the case for the "render-focused-layout", for example, in which the secondary splitter contains three widgets.

These commands should be revised to allow for the new layout possibilities and how they nest their widgets.

For layouts defined by a layout structure - almost all of them - it's easy to tell what's in the secondary and main splitters. A layout structure includes an ordered dictionary defined by tuples (widget, splitter), like this:

```
    'SPLITTERS': OrderedDict(
            (('outlineFrame', 'secondary_splitter'),
            ('bodyFrame', 'secondary_splitter'),
            ('logFrame', 'secondary_splitter'),
            (VRX_PLACEHOLDER_NAME, 'body-vr-splitter'),
            ('secondary_splitter', 'main_splitter'),
            ('body-vr-splitter', 'main_splitter'))
        ),
```

For this layout, there are three widgets that share the secondary splitter, in the order listed, and two widgets in the main splitter.

**Comment by edreamleo** [comment]
On Thu, Oct 24, 2024 at 10:36 AM Thomas Passin ***@***.***>
wrote:

> I think these commands assume that the log pane shares the secondary
> splitter with some other widget, like the outline frame. Some of the new
> layouts may not meet this assumption. Or the commands may act differently
> for different layouts.
>
...

> These commands should be revised to allow for the new layout possibilities
> and how they nest their widgets.
>
For layouts defined by a layout structure - almost all of them - it's easy
> to tell what's in the secondary and main splitters.
>
Thanks for this hint. I'll look into this issue.

Edward

**Comment by tbpassin** [comment]
Adding to the store of diagnostic information, with the "vertical-thirds2" layout, "contract-log-pane" and "expand-log-pane" work in reverse.

**Comment by tbpassin** [comment]
One algorithm for adjusting the log size could be this:

find log frame `parent()` # a splitter
`splitter.count()` - number of widgets in this splitter
'splitter.index(log_widget)` - position of the log widget within the splitter
`splitter.sizes()` - current sizes of splitter widgets
recompute sizes
`splitter.setSizes(new_sizes))`

**Comment by edreamleo** [comment]
@tbpassin Thanks.

**Comment by edreamleo** [comment]
@tbpassin Your suggested algorithm is on the right track because it is dynamic. That is, the algorithm will work regardless of which layout created the present configuration of widgets.

I appreciate your ongoing work, but imo merging PR #4121 should have top priority. Thanks.

**Comment by tbpassin** [comment]
> imo merging PR #4121 should have top priority.

I agree. Coming shortly.

**Comment by edreamleo** [comment]
@tbpassin I'm going to tackle this issue. I'll ask you for help if necessary.

**Comment by tbpassin** [comment]
All right.

**Comment by edreamleo** [comment]
@tbpassin Thanks. I'll keep you informed. I've just created PR #4136.

The first commit involves infrastructure. All commands delegate their work to the `LayoutCacheWidget` class.

**Comment by edreamleo** [comment]
Closed via PR #4136.
